File: python/leetcode/042.py
class Solution:
    def trap(self, height):
        l , r = 0, len(height) - 1
        total, pillar, h = 0, 0, 0
        while l <= r:
            tmp = min(height[l], height[r])
            total += (r - l + 1) * (tmp - h) if tmp > h else 0
            pillar += tmp
            h = max(h, tmp)
            if height[l] == tmp:
                l += 1
            else:
                r -= 1
        return total - pillar

File: cpp/leetcode/042.cpp
class Solution {
  public:
    int trap(vector<int> &height) {
        int total = 0, pill = 0, h = 0;
        int left = 0, right = height.size() - 1;
        while (left <= right) {
            int tmp = min(height[left], height[right]);
            total += tmp > h ? (right - left + 1) * (tmp - h) : 0;
            h = max(h, tmp);
            pill += height[left] < height[right] ? height[left++] : height[right--];
        }
        return total - pill;
    }
};